README.md
![demo](https://raw.githubusercontent.com/guanyuxin/baogame/master/doc/demo1.gif)
#在线游戏
访问： http://guanyuxin.com/

由于官方服务器带宽和延迟较大，想玩的同学可以通过搭建私服在局域网中和好朋友一起玩。搭建私服方法在末尾给出。
<br>
*2016-01-17修改了手雷的控制方式：按住act扔出，按住的时间越久，扔的越远，扔出后一段时间爆炸；蹲下时扔出的手雷会贴着地滚。*

--
#游戏简介

###游戏目标:
	把其他人推下水
	使用道具消灭其他人
	have fun

###加入:
	通过浏览器打开后可以查看目前的游戏状态，选择加入后可以进行游戏
	
	加入前给自己起个有个性的名字吧
	
	自己的角色顶部有黄色名称，敌人角色顶部有红色名称

###移动碰撞：
	wasd控制移动，当玩家处于平台上时，w为跳跃，d为下蹲；当玩家处于梯子附近时（头顶出现上下箭头），w,d为爬梯子上下
	
    ** 手机上使用虚拟按键控制 **

	两个玩家接触后会产生碰撞，将两个玩家弹开，使用这个机制把敌人推下平台吧
	
	通常情况下，两个玩家碰撞时，跳起或者蹲下的一方会有优势

###道具：
	游戏中有紫色的能量球，玩家吃到后会产生各种能力或者效果，有些道具的效果能力强大，好好使用他们。详情在道具部分介绍。

![demo](https://raw.githubusercontent.com/guanyuxin/baogame/master/doc/demo2.gif)

#道具
###毒药:
![drug](https://raw.githubusercontent.com/guanyuxin/baogame/master/doc/drug.png)

*大部分道具是强大而有益的，但是看到毒药你还是离他远一些为好，他会让吃到他的玩家立即死亡*

###手枪:
![drug](https://raw.githubusercontent.com/guanyuxin/baogame/master/doc/gun.png)

*按q向面前发射一颗子弹，消灭任何敢于正面对抗你的敌人，注意：只有三发子弹，请节约使用。无法消灭下蹲或者跳起的敌人*

###无敌:
![drug](https://raw.githubusercontent.com/guanyuxin/baogame/master/doc/power.png)

*并不是真正的无敌，但是会让你直接消灭所有敢于触碰你的敌人，并且他们无法给你造成碰撞*

###隐身:
![drug](https://raw.githubusercontent.com/guanyuxin/baogame/master/doc/hide.png)

*使用后慢慢从你的敌人视野里面消失，谁能和看不见的敌人战斗呢？*

###惊喜:
![drug](https://raw.githubusercontent.com/guanyuxin/baogame/master/doc/random.png)

*surprise !*

###喷气背包:
![drug](https://raw.githubusercontent.com/guanyuxin/baogame/master/doc/flypack.png)

*跳的不够高？干嘛不飞呢！跳起后再次按w进入飞行模式，让那些只会蹦跶的人羡慕吧。等等,好像没油了...*

###手雷:
![drug](https://raw.githubusercontent.com/guanyuxin/baogame/master/doc/grenade.png)

*按住act扔出，按住的时间越久，扔的越远，扔出后一段时间爆炸；蹲下时扔出的手雷会贴着地滚。手雷扔出的速度受角色影响，所以跑动时手雷速度更快，跳起时仍的更高*

--

![demo](https://raw.githubusercontent.com/guanyuxin/baogame/master/doc/demo3.gif)

#搭建私服方法--通过npm [稳定版]

1.安装node4.2.4(或者以上版本)和npm
# **如果安装出现问题请尝试将node升级至最新版本

2.shell中执行以下代码：

```
npm install fuzion-game &&
cd node_modules/fuzion-game/ &&
node app.js
```

3.打开http://localhost:8030  就可以开始玩了

4.把localhost替换成你的域名或者ip，然后分享给你的朋友，一起玩吧

--

#搭建私服方法--使用github [最新版]

将上面方法的第二部替换为：

```
git clone https://github.com/guanyuxin/baogame
cd baogame
npm install
node app.js
```
--

#服务器管理

```
#启动参数：
node app.js [port=端口，默认8030] [code=管理员口令，默认admin] [room=房间数目，默认1]
```
http://localhost:port/admin  可以进入管理界面，需要localStorage中设置code=管理员口令，然后可以创建物品或者封禁用户ip


changelog.md
# 1.2.0 2016-01-28 
增加了单人教学

# 1.1.0 2016-01-17
增加了手雷  按住act扔出，按住的时间越久，扔的越远，扔出后一段时间爆炸；蹲下时扔出的手雷会贴着地滚。

# 1.0.0 2016-01-15
初始版本


game.js
var Map = require('./map.js');
var User = require('./user.js');
var Item = require('./item.js');
var Con = require('./con.js');


function userCollide(a, b, game) {
	//不碰撞情况
	if (a.dead || b.dead) {return}
	if((a.x-b.x)*(a.x-b.x) + (a.y-b.y)*(a.y-b.y) > game.props.userWidth*game.props.userWidth) {return;}

	//带电情况
	if (a.carry == "power" && b.carry != "power") {
		b.killed('power', a);
		b.vx = (b.x - a.x)/2;
		if (b.carry == "bomb") {
			a.carry = "bomb";
			a.carryCount = b.carryCount;
			b.carry = '';
		}
		return;
	} else if (a.carry != "power" && b.carry == "power") {
		a.killed('power', b);
		a.vx = (a.x - b.x)/2;
		if (a.carry == "bomb") {
			b.carry = "bomb";
			b.carryCount = b.carryCount;
			a.carry = '';
		}
		return;
	} else if (a.carry == "power" && b.carry == "power") {
		a.carry = '';
		b.carry = '';
	}
	//排除刚刚碰撞
	if (a.ignore[b.id] > 0 || b.ignore[a.id] > 0) {return}
	
	if (b.carry == "bomb" && a.carry != "bomb") {
		a.carry = "bomb";
		a.carryCount = b.carryCount;
		b.carry = '';
	} else if (a.carry == "bomb" && b.carry != "bomb") {
		b.carry = "bomb";
		b.carryCount = a.carryCount;
		a.carry = '';
	}
	//正常情况
	if (a.onFloor && b.onFloor) {
		if (a.crawl && !b.crawl) {
			b.vy = 5;
			b.danger = true;
		} else if (!a.crawl && b.crawl) {
			a.vy = 5;
			a.danger = true;
		} else {
			if (a.crawl && b.crawl) {
				a.crawl = false;
				b.crawl = false;
			}
			var tmp = a.vx;
			a.vx = b.vx;
			b.vx = tmp;
			
			a.vy = 2.5;
			b.vy = 2.5;
		}
	} else if (a.onFloor && !b.onFloor) {
		if (a.crawl) {
			a.vx = b.vx / 2;
			b.vx = -b.vx / 2;
			a.vy = 2.5;
			b.vy = 2.5;
		} else {
			a.vx = b.vx;
			b.vx /= 2;
			a.vy = 2.5;
			a.danger = true;
		}
	} else if (!a.onFloor && b.onFloor) {
		if (b.crawl) {
			b.vx = a.vx / 2;
			a.vx = -a.vx / 2;
			b.vy = 2.5;
			a.vy = 2.5;
		} else {
			b.vx = a.vx;
			a.vx /= 2;
			b.vy = 2.5;
			b.danger = true;
		}
	} else {
		var tmp = a.vx;
		a.vx = b.vx;
		b.vx = tmp;
		a.danger = true;
		b.danger = true;
	}
	//自然抗拒
	if (a.x < b.x) {
		if (!a.crawl) {
			a.vx -= 1;
		}
		if (!b.crawl) {
			b.vx += 1;
		}
	} else {
		if (!a.crawl) {
			a.vx += 1;
		}
		if (!b.crawl) {
			b.vx -= 1;
		}
	}
	//阻止近期碰撞
	a.ignore[b.id] = 40;
	b.ignore[a.id] = 40;
	a.fireing = false;
	b.fireing = false;
	a.mining = false;
	b.mining = false;
	a.lastTouch = b.id;
	b.lastTouch = a.id;
}

function eatItem (a, b, game) {
	if (a.dead || b.dead) {return}
	if (a.carry == "bomb") {return}
	if((a.x-b.x)*(a.x-b.x) + (a.y+game.props.userHeight/2-b.y)*(a.y+game.props.userHeight/2-b.y) > (game.props.userWidth+game.props.itemSize)*(game.props.userWidth+game.props.itemSize)/4) {return;}
	b.dead = true;
	if (b.type == 'gun') {
		a.carry = 'gun';
		a.carryCount = 3;
	} else if (b.type == 'drug') {
		a.killed('drug');
	} else if (b.type == 'power') {
		a.carry = "power";
		a.carryCount = 1000;
	} else if (b.type == 'mine') {
		a.carry = 'mine';
		a.carryCount = 2;
	} else if (b.type == 'hide') {
		a.carry = 'hide';
		a.carryCount = 1000;
	} else if (b.type == 'random') {
		a.carry = 'bomb';
		a.carryCount = 600;
	} else if (b.type == 'flypack') {
		a.carry = 'flypack';
		a.carryCount = 200;
	}
}

var Game = function () {
	this.props = {
		w: 1100,
		h: 600,
		blockWidth: 50,
		blockHeight: 70,
		userHeight: 40,
		userWidth: 40,
		itemSize: 15,
		tileW: 22,
		tileH: 8
	}

	this.users = [];
	this.cons = [];
	this.items = [];
	this.bodies = [];
	this.mines = [];
	this.map = new Map(this, 22, 8);
	this.tick = 0;
	var _this = this;
	this.runningTimer = setInterval(function () {
		_this.update();
	}, 17);
}
//增加玩家
Game.prototype.addUser = function (name) {
	var u = new User(this, name);
	var place = this.map.born();
	u.x = place.x;
	u.y = place.y + this.props.blockHeight/2;
	this.users.push(u);
	return u;
}
//获得玩家（或者尸体）
Game.prototype.getUser = function (uid) {
	for (var i = 0; i < this.users.length; i++) {
		if (this.users[i].id == uid) {
			return this.users[i];
		}
	}
	for (var i = 0; i < this.bodies.length; i++) {
		if (this.bodies[i].id == uid) {
			return this.bodies[i];
		}
	}
}

//发生爆炸了
Game.prototype.explode = function (x, y , byUser) {
	var _this = this;
	this.users.forEach(function (user) {
		var ux = user.x;
		var uy = user.y + _this.props.userHeight;
		var dist = (ux - x)*(ux - x) + (uy - y)*(uy - y);
		if (dist < 10000) {
			user.killed('bomb', byUser);
		}
		if (dist < 22500) {
			var r = Math.atan2(uy - y, ux - x);
			var force = 45000 / (dist + 2500);
			user.vx += force * Math.cos(r);
			user.vy += force * Math.sin(r);
			user.danger = true;
		} 
	});
	this.announce('explode', {x:x,y:y});
}

//发生枪击
Game.prototype.checkShot = function (u) {
	var game = this;
	var x = u.x;
	var y = u.y + game.props.userHeight/2;
	var f = u.faceing;

	this.users.forEach(function (user) {
		if (!user.crawl && f < 0 && x > user.x && user.y <= y && user.y + game.props.userHeight >= y) {
			user.killed('gun', u);
			user.vx = 6 * f;
		}

		if (!user.crawl && f > 0 && x < user.x && user.y <= y && user.y + game.props.userHeight >= y) {
			user.killed('gun', u);
			user.vx = 6 * f;
		}
	});
}
Game.prototype.award = function (u) {
	u.score++;
	if (u.score == 5 && !u.dead) {
		this.announce('winner', u.getData());
	}
}
Game.prototype.addMine = function (user) {
	var x = user.x + user.faceing * 40;
	if (this.map.onFloor(x, user.y)) {
		this.mines.push({
			x: x,
			y: user.y,
			creater: user,	
		});
		return true;
	}
	return false;
}
Game.prototype.checkMine = function (user) {
	for (var i = this.mines.length - 1; i >= 0; i--) {
		var mine = this.mines[i];
		if (Math.abs(user.x - mine.x) < 10 && Math.abs(user.y - mine.y) < 5) {
			user.killed('mine', mine.creater);
			mine.dead = true;
			return true;
		}
	}
	return false;
}
//链接
Game.prototype.addCon = function (soc) {
	this.cons.push(new Con(soc, this));
}
//链接关闭
Game.removeCon = function (con) {

}
//分发事件
Game.prototype.announce = function (type, data) {
	for (var i = 0; i < this.cons.length; i++) {
		this.cons[i].soc.emit(type, data);
	}
}

//游戏主流程
Game.prototype.update = function () {
	this.tick++;
	//生成物品（如果需要）
	if (this.items.length < this.users.length && Math.random() * 500 < this.users.length) {
		this.items.push(new Item(this));
	}
	//物品更新
	for(var i = this.items.length - 1; i >= 0; i--) {
		var item = this.items[i];
		item.update();
	};
	//碰撞检测
	for (var i = 0; i < this.users.length; i++) {
		for (var j = i + 1; j < this.users.length; j++) {
			userCollide(this.users[i], this.users[j], this);
		}
		for (var j = 0; j < this.items.length; j++) {
			eatItem(this.users[i], this.items[j], this);
		}
	}
	//user更新
	this.users.forEach(function (user) {
		user.update();
	});
	//分发状态
	this.sendTick();
	//清理死亡的人物/物品
	this.clean();
}
Game.prototype.clean = function () {
	for(var i = this.items.length - 1; i >= 0; i--) {
		var item = this.items[i];
		if (!item.dead) {
		} else {
			this.items.splice(i, 1);
		}
	};
	for(var i = this.mines.length - 1; i >= 0; i--) {
		var mine = this.mines[i];
		if (!mine.dead) {
		} else {
			this.mines.splice(i, 1);
		}
	};
	for(var i = this.users.length - 1; i >= 0; i--) {
		var user = this.users[i];
		if (!user.dead) {
		} else {
			this.users.splice(i, 1);
			this.bodies.push(user);
		}
	};
}
Game.prototype.sendTick = function () {
	
	var itemdata = [];
	for (var i = 0; i < this.items.length; i++) {
		itemdata.push(this.items[i].getData());
	}
	var userdata = [];
	this.users.forEach(function (user) {
		userdata.push(user.getData());
	});
	for (var i = 0; i < this.cons.length; i++) {
		var mines = [];
		var p1 = this.cons[i].p1 && this.cons[i].p1.id;
		var p2 = this.cons[i].p2 && this.cons[i].p2.id;
		this.mines.forEach(function(mine) {
			if ((mine.creater.id == p1 && !p2) || mine.dead) {
				mines.push({
					x: mine.x,
					y: mine.y,
					dead: mine.dead
				});
			}
		});
		
		this.cons[i].soc.emit('tick', {
			users: userdata,
			items: itemdata,
			mines: mines,
			p1: p1,
			p2: p2
		});
	}
}
module.exports = Game;